// ngx-barcode-scanner/esm5/lib/barcode-scanner-livestream-overlay/barcode-scanner-livestream-overlay.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild } from '@angular/core';
import { BarecodeScannerLivestreamComponent } from '../barcode-scanner-livestream/barcode-scanner-livestream.component';
var BarecodeScannerLivestreamOverlayComponent = /** @class */ (function () {
    function BarecodeScannerLivestreamOverlayComponent() {
        this.valueChanges = new EventEmitter();
        this._showScanner = false;
    }
    Object.defineProperty(BarecodeScannerLivestreamOverlayComponent.prototype, "showScanner", {
        get: /**
         * @return {?}
         */
        function () {
            return this._showScanner;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    BarecodeScannerLivestreamOverlayComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.scanner.stop();
    };
    /**
     * @return {?}
     */
    BarecodeScannerLivestreamOverlayComponent.prototype.show = /**
     * @return {?}
     */
    function () {
        this._showScanner = true;
        this.scanner.start();
    };
    /**
     * @return {?}
     */
    BarecodeScannerLivestreamOverlayComponent.prototype.hide = /**
     * @return {?}
     */
    function () {
        this._showScanner = false;
        this.scanner.stop();
    };
    /**
     * @param {?} result
     * @return {?}
     */
    BarecodeScannerLivestreamOverlayComponent.prototype.onValueChanges = /**
     * @param {?} result
     * @return {?}
     */
    function (result) {
        this.valueChanges.next(result);
    };
    BarecodeScannerLivestreamOverlayComponent.decorators = [
        { type: Component, args: [{
                    selector: 'barcode-scanner-livestream-overlay',
                    template: "<div class=\"barcode-scanner-livestream-overlay\" [hidden]=\"!showScanner\">\n    <div class=\"barcode-scanner-livestream-overlay-content\">\n        <div class=\"barcode-scanner-livestream-overlay-close\" (click)=\"hide()\">X</div>\n        <barcode-scanner-livestream [type]=\"type\" (valueChanges)=\"onValueChanges($event)\"></barcode-scanner-livestream>\n    </div>\n</div>",
                    styles: [".barcode-scanner-livestream-overlay{overflow:hidden;position:fixed;top:0;bottom:0;left:0;right:0;width:100%;background-color:rgba(0,0,0,.3);z-index:1000}.barcode-scanner-livestream-overlay .barcode-scanner-livestream-overlay-content{top:50%;position:absolute;left:50%;-webkit-transform:translate(-50%,-50%);transform:translate(-50%,-50%);width:90%;max-height:90%;max-width:800px}.barcode-scanner-livestream-overlay .barcode-scanner-livestream-overlay-content .barcode-scanner-livestream-overlay-close{position:absolute;right:0;padding:.5rem;width:2rem;height:2rem;line-height:2rem;text-align:center;background-color:#fff;cursor:pointer;border:3px solid #000;font-size:1.5rem;margin:-1rem;border-radius:2rem;z-index:100;box-sizing:content-box}"]
                }] }
    ];
    BarecodeScannerLivestreamOverlayComponent.propDecorators = {
        type: [{ type: Input }],
        valueChanges: [{ type: Output }],
        scanner: [{ type: ViewChild, args: [BarecodeScannerLivestreamComponent,] }]
    };
    return BarecodeScannerLivestreamOverlayComponent;
}());
export { BarecodeScannerLivestreamOverlayComponent };
if (false) {
    /** @type {?} */
    BarecodeScannerLivestreamOverlayComponent.prototype.type;
    /** @type {?} */
    BarecodeScannerLivestreamOverlayComponent.prototype.valueChanges;
    /** @type {?} */
    BarecodeScannerLivestreamOverlayComponent.prototype.scanner;
    /** @type {?} */
    BarecodeScannerLivestreamOverlayComponent.prototype._showScanner;
}
