// ngx-barcode-scanner/fesm2015/ngx-barcode-scanner.js
import { onProcessed, onDetected, init, start, stop, canvas, ImageDebug } from 'quagga';
import { CommonModule } from '@angular/common';
import { Component, Input, Output, EventEmitter, ViewChild, ViewEncapsulation, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @param {?} value
 * @return {?}
 */
function mapToReader(value) {
    return [value + "_reader"];
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
const DEFAULT_CONFIG = {
    inputStream: {
        name: 'Live',
        type: 'LiveStream',
        target: null,
        constraints: {
            width: { min: 640 },
            height: { min: 480 },
            aspectRatio: { min: 1, max: 100 },
            facingMode: 'environment',
        },
        singleChannel: false // true: only the red color-channel is read
    },
    locator: {
        patchSize: 'medium',
        halfSample: true
    },
    locate: true,
    numOfWorkers: 4,
    decoder: {
        readers: ['code_128_reader']
    }
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class BarecodeScannerLivestreamComponent {
    constructor() {
        // Outputs
        this.valueChanges = new EventEmitter();
        this.started = false;
        this.configQuagga = DEFAULT_CONFIG;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.stop();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        this.retart();
    }
    /**
     * @return {?}
     */
    _init() {
        return new Promise((resolve, reject) => {
            onProcessed((result) => this.onProcessed(result));
            onDetected((result) => this.onDetected(result));
            this.configQuagga.inputStream.target = this.barecodeScanner.nativeElement;
            if (this.type) {
                this.configQuagga.decoder.readers = mapToReader(this.type);
            }
            init(this.configQuagga, (err) => {
                if (err) {
                    console.log(err);
                    return reject(err);
                }
                resolve();
            });
        });
    }
    /**
     * @return {?}
     */
    start() {
        if (!this.started) {
            return this._init().then(() => {
                start();
                this.started = true;
                console.log('started');
            });
        }
        return Promise.resolve();
    }
    /**
     * @return {?}
     */
    stop() {
        if (this.started) {
            stop();
            this.started = false;
            console.log('stopped');
        }
    }
    /**
     * @return {?}
     */
    retart() {
        if (this.started) {
            this.stop();
            this.start();
        }
    }
    /**
     * @return {?}
     */
    isStarted() {
        return this.started;
    }
    /**
     * @param {?} result
     * @return {?}
     */
    onProcessed(result) {
        /** @type {?} */
        let drawingCtx = canvas.ctx.overlay;
        /** @type {?} */
        let drawingCanvas = canvas.dom.overlay;
        if (result) {
            if (result.boxes) {
                drawingCtx.clearRect(0, 0, parseInt(drawingCanvas.getAttribute('width')), parseInt(drawingCanvas.getAttribute('height')));
                result.boxes.filter(function (box) {
                    return box !== result.box;
                }).forEach(function (box) {
                    ImageDebug.drawPath(box, {
                        x: 0,
                        y: 1,
                    }, drawingCtx, {
                        color: 'green',
                        lineWidth: 2,
                    });
                });
            }
            if (result.box) {
                ImageDebug.drawPath(result.box, {
                    x: 0,
                    y: 1,
                }, drawingCtx, {
                    color: '#00F',
                    lineWidth: 2,
                });
            }
            if (result.codeResult && result.codeResult.code) {
                ImageDebug.drawPath(result.line, {
                    x: 'x',
                    y: 'y',
                }, drawingCtx, {
                    color: 'red',
                    lineWidth: 3,
                });
            }
        }
    }
    /**
     * @param {?} result
     * @return {?}
     */
    onDetected(result) {
        this.valueChanges.next(result);
    }
}
BarecodeScannerLivestreamComponent.decorators = [
    { type: Component, args: [{
                selector: 'barcode-scanner-livestream',
                template: "<div #BarecodeScanner class=\"scanner\" [hidden]=\"!isStarted()\">\n</div>",
                encapsulation: ViewEncapsulation.None,
                styles: [".scanner{position:relative}.scanner canvas,.scanner video{max-width:100%;width:100%}.scanner canvas.drawingBuffer{position:absolute;left:0;top:0}"]
            }] }
];
BarecodeScannerLivestreamComponent.propDecorators = {
    type: [{ type: Input }],
    valueChanges: [{ type: Output }],
    barecodeScanner: [{ type: ViewChild, args: ['BarecodeScanner',] }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class BarecodeScannerLivestreamModule {
}
BarecodeScannerLivestreamModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    BarecodeScannerLivestreamComponent
                ],
                exports: [
                    BarecodeScannerLivestreamComponent
                ]
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class BarecodeScannerLivestreamOverlayComponent {
    constructor() {
        this.valueChanges = new EventEmitter();
        this._showScanner = false;
    }
    /**
     * @return {?}
     */
    get showScanner() {
        return this._showScanner;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.scanner.stop();
    }
    /**
     * @return {?}
     */
    show() {
        this._showScanner = true;
        this.scanner.start();
    }
    /**
     * @return {?}
     */
    hide() {
        this._showScanner = false;
        this.scanner.stop();
    }
    /**
     * @param {?} result
     * @return {?}
     */
    onValueChanges(result) {
        this.valueChanges.next(result);
    }
}
BarecodeScannerLivestreamOverlayComponent.decorators = [
    { type: Component, args: [{
                selector: 'barcode-scanner-livestream-overlay',
                template: "<div class=\"barcode-scanner-livestream-overlay\" [hidden]=\"!showScanner\">\n    <div class=\"barcode-scanner-livestream-overlay-content\">\n        <div class=\"barcode-scanner-livestream-overlay-close\" (click)=\"hide()\">X</div>\n        <barcode-scanner-livestream [type]=\"type\" (valueChanges)=\"onValueChanges($event)\"></barcode-scanner-livestream>\n    </div>\n</div>",
                styles: [".barcode-scanner-livestream-overlay{overflow:hidden;position:fixed;top:0;bottom:0;left:0;right:0;width:100%;background-color:rgba(0,0,0,.3);z-index:1000}.barcode-scanner-livestream-overlay .barcode-scanner-livestream-overlay-content{top:50%;position:absolute;left:50%;-webkit-transform:translate(-50%,-50%);transform:translate(-50%,-50%);width:90%;max-height:90%;max-width:800px}.barcode-scanner-livestream-overlay .barcode-scanner-livestream-overlay-content .barcode-scanner-livestream-overlay-close{position:absolute;right:0;padding:.5rem;width:2rem;height:2rem;line-height:2rem;text-align:center;background-color:#fff;cursor:pointer;border:3px solid #000;font-size:1.5rem;margin:-1rem;border-radius:2rem;z-index:100;box-sizing:content-box}"]
            }] }
];
BarecodeScannerLivestreamOverlayComponent.propDecorators = {
    type: [{ type: Input }],
    valueChanges: [{ type: Output }],
    scanner: [{ type: ViewChild, args: [BarecodeScannerLivestreamComponent,] }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class BarecodeScannerLivestreamOverlayModule {
}
BarecodeScannerLivestreamOverlayModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    BarecodeScannerLivestreamModule
                ],
                declarations: [
                    BarecodeScannerLivestreamOverlayComponent
                ],
                exports: [
                    BarecodeScannerLivestreamOverlayComponent
                ]
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { BarecodeScannerLivestreamModule, BarecodeScannerLivestreamComponent, BarecodeScannerLivestreamOverlayModule, BarecodeScannerLivestreamOverlayComponent };
